### PR TITLE
✅ Add integration tests for SXG build.

### PR DESCRIPTION
**Pull request by xiexr151e**
This PR piggybacks off #28336. It enables integration testing for the SXG build.
This PR also unblocks #30126.

### CODE REVIEW

File: build-system/pr-check/sxg-tests.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+const timedExecOrDie = (cmd, unusedFileName) =>
+  timedExecOrDieBase(cmd, FILENAME);
+
+function main() {
+  const startTime = startTimer(FILENAME, FILENAME);
+
+  if (!isTravisPullRequestBuild()) {
+    timedExecOrDie('gulp update-packages');
+    timedExecOrDie('gulp dist --fortesting --sxg');
+    timedExecOrDie('gulp integration --nobuild --compiled --headless');
```
Comment (rsimha) [inline build-system/pr-check/sxg-tests.js:46]: I'm pretty sure this command also needs an `--sxg` flag in order to have the test server launch in the correct mode.

File: build-system/pr-check/sxg-tests.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    printChangeSummary(FILENAME);
+    const buildTargets = determineBuildTargets(FILENAME);
+    if (
+      buildTargets.has('RUNTIME') ||
+      buildTargets.has('FLAG_CONFIG') ||
+      buildTargets.has('INTEGRATION_TEST')
+    ) {
+      downloadSxgDistOutput(FILENAME);
+      timedExecOrDie('gulp update-packages');
+      timedExecOrDie('gulp integration --nobuild --compiled --headless');
```
Comment (rsimha) [inline build-system/pr-check/sxg-tests.js:57]: Same here.

File: build-system/pr-check/utils.js [hunk 3]
```diff
@@ -363,6 +374,7 @@ module.exports = {
   downloadBuildOutput,
   downloadDistOutput,
   downloadEsmDistOutput,
+  downloadSxgDistOutput,
```
Comment (rsimha) [inline build-system/pr-check/utils.js:377]: This PR isn't building and uploading the SXG binaries. You'd need to implement `uploadSxgDistOutput` as well, and add a new parallel job to the build stage. You can use the `Module Dist, Bundle Size` job as an example, except I don't think it's necessary to keep track of the bundle size of SXG builds, so you could just call it `SXG Dist`.

**Review by rsimha — CHANGES_REQUESTED** [review]
Thanks for working on this. The PR is on the right track, but is missing a few critical pieces, so I've put in a blocking review for now. Happy to take another look.

File: build-system/pr-check/sxg-tests.js (hunk 1 above)
Comment (xiexr151e) [reply build-system/pr-check/sxg-tests.js:46]: `update-packages` or `integration`?

File: build-system/pr-check/utils.js (hunk 3 above)
Comment (xiexr151e) [reply build-system/pr-check/utils.js:386]: Please take a look at the updated `sxg-tests.js`. I tried to mimic the `module-dist-bundle-size.js` task, but I'm not 100% sure if it is the correct way to do it.

Comment (rsimha) [reply build-system/pr-check/utils.js:386]: You'll need to upload the sxg output in a new parallel job that's run in the build stage on Travis, not the test stage. Best way is to make a copy of the `Module Dist, Bundle Size` job and call it `SXG Dist` (skip the bundle size part for now). @erwinmombay can help you with this.

Comment (erwinmombay) [reply build-system/pr-check/utils.js:386]: @rsimha will do. have a meeting with @xiexr151e in a bit so we'll address it. thank you

Comment (xiexr151e) [reply build-system/pr-check/utils.js:386]: I made a new parallel job in the `.travis.yml`, as well as the corresponding `sxg-dist.js`. Please take a look at it.

File: build-system/pr-check/sxg-tests.js (hunk 1 above)
Comment (xiexr151e) [reply build-system/pr-check/sxg-tests.js]: I just added an `sxg` flag to the integration task. Please take a look.

File: build-system/pr-check/sxg-dist.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+/**
+ * @fileoverview
+ * This script builds the esm minified AMP runtime and runs the bundle size check.
+ * This is run during the CI stage = build; job = module dist, bundle size.
```
Comment (rsimha) [inline build-system/pr-check/sxg-dist.js:21]: Update these comments.

File: build-system/pr-check/sxg-dist.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  stopTimer,
+  stopTimedJob,
+  timedExecOrDie: timedExecOrDieBase,
+  uploadSxgDistOutput,
+} = require('./utils');
+const {determineBuildTargets} = require('./build-targets');
+const {isTravisPullRequestBuild} = require('../common/travis');
+const {runYarnChecks} = require('./yarn-checks');
+
+const FILENAME = 'module-dist-bundle-size.js';
```
Comment (rsimha) [inline build-system/pr-check/sxg-dist.js:37]: Update this.

File: build-system/pr-check/sxg-tests.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+const FILELOGPREFIX = colors.bold(colors.yellow(`${FILENAME}:`));
+const timedExecOrDie = (cmd, unusedFileName) =>
+  timedExecOrDieBase(cmd, FILENAME);
+
+function main() {
+  const startTime = startTimer(FILENAME, FILENAME);
+
+  if (!isTravisPullRequestBuild()) {
+    timedExecOrDie('gulp update-packages');
+    timedExecOrDie('gulp dist --fortesting --sxg');
```
Comment (rsimha) [inline build-system/pr-check/sxg-tests.js:45]: Don't do a build here. Instead, download the build artifacts that were uploaded during the previous stage. (See the visual tests or the E2E tests for an example.)

File: build-system/pr-check/utils.js (hunk 3 above)
Comment (rsimha) [reply build-system/pr-check/utils.js:386]: Reviewed and posted some new comments. Getting closer!

File: build-system/pr-check/sxg-tests.js (hunk 6 above)
Comment (xiexr151e) [reply build-system/pr-check/sxg-tests.js:45]: So within this condition, I would just do a single download for the sxg dist output, right?

Comment (xiexr151e) [reply build-system/pr-check/sxg-tests.js]: Actually, I think I got things mixed up.
Please take another look at `sxg-tests` and `sxg-dist`. I believe it should be correct now.

File: build-system/tasks/report-test-status.js [hunk 7]
```diff
@@ -32,7 +32,7 @@ const IS_LOCAL_CHANGES = !!argv.local_changes;
 const IS_DIST = !!argv.compiled;
 
 const TEST_TYPE_SUBTYPES = new Map([
-  ['integration', ['local', 'minified']],
+  ['integration', ['local', 'minified', 'sxg']],
```
Comment (rsimha) [inline build-system/tasks/report-test-status.js:35]: Thanks for adding this. There are a couple more changes needed to get this to work:

a. Add a new const called `IS_SXG` above `IS_DIST` (line 32), based on `argv.sxg`.
b. In `inferTestType()`, add a new `else if` block before `IS_DIST` which returns `${type}/sxg`.

File: build-system/pr-check/sxg-tests.js (hunk 6 above)
Comment (rsimha) [reply build-system/pr-check/sxg-tests.js]: The new files look good. I've added a new comment to complete the test reporting stuff, after which this should be good to go. Thanks!

File: build-system/tasks/report-test-status.js (hunk 7 above)
Comment (xiexr151e) [reply build-system/tasks/report-test-status.js:36]: Done.

**Comment by rsimha** [comment]
Another step closer! The SXG build was [successful](https://travis-ci.org/github/ampproject/amphtml/jobs/730336173#L639), but it looks like most of the integration tests [are failing](https://travis-ci.org/github/ampproject/amphtml/jobs/730336178#L2583). Any idea what's missing? Perhaps you need to do some additional setup, similar to ESM?

Here's how ESM sets up the server mode:

https://github.com/ampproject/amphtml/blob/ebaa3771bd7a8393f23a4a68657e9218934543cf/build-system/server/app-utils.js#L31-L75

/cc @erwinmombay

**Pull request closed without merging**